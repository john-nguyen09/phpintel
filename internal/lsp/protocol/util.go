package protocol

import (
	"strconv"
	"strings"
)

// IsInRange compare whether a position is within a range
func IsInRange(position Position, theRange Range) int {
	start := theRange.Start
	end := theRange.End

	if position.Line < start.Line || (position.Line == start.Line && position.Character < start.Character) {
		return -1
	}
	if position.Line > end.Line || (position.Line == end.Line && position.Character >= end.Character) {
		if ComparePos(start, end) == 0 {
			return -1
		}
		return 1
	}
	return 0
}

// ComparePos returns -1, 0 or 1 indicating whether a is before, equal to or after b
func ComparePos(a Position, b Position) int {
	if a.Line < b.Line {
		return -1
	}
	if a.Line == b.Line {
		if a.Character < b.Character {
			return -1
		}
		if a.Character == b.Character {
			return 0
		}
		if a.Character > b.Character {
			return 1
		}
	}
	return 1
}

// CompareRange compares 2 ranges
func CompareRange(a Range, b Range) int {
	if IsInRange(a.End, b) < 0 {
		return -1
	}
	if IsInRange(a.Start, b) > 0 {
		return 1
	}
	return 0
}

// PosFromString creates Position from string
func PosFromString(str string) Position {
	values := strings.Split(str, ":")
	line, err := strconv.Atoi(values[0])
	if err != nil {
		panic(err)
	}
	character, err := strconv.Atoi(values[1])
	if err != nil {
		panic(err)
	}
	return Position{
		Line:      line,
		Character: character,
	}
}

// RangeFromString creates Range from string
func RangeFromString(str string) Range {
	values := strings.Split(str, "-")
	start := PosFromString(values[0])
	end := PosFromString(values[1])
	return Range{
		Start: start,
		End:   end,
	}
}
