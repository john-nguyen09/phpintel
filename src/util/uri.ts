import URI from 'vscode-uri/lib/umd';
import * as path from 'path';

export function pathToUri(filePath: string): string {
    filePath = path.resolve(filePath).replace(/\\/g, '/');

    if (filePath[0] != '/') {
        filePath = '/' + filePath;
    }
    
    return encodeURI('file://' + filePath);
}

export function uriToPath(uri: string) {
    return URI.parse(uri).fsPath;
}

export function toRelative(uri: string) {
    let baseUri = pathToUri(path.resolve(__dirname, '..' , '..'));

    if (uri.indexOf(baseUri) === 0) {
        return uri.substr(baseUri.length + 1);
    }

    return uri;
}