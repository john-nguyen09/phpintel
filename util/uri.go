package util

import (
	"crypto/md5"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"net/url"
	"strings"
)

// URIToPath convert URI to file path
func URIToPath(uri string) (string, error) {
	urlIns, err := url.ParseRequestURI(uri)

	if err != nil {
		return "", err
	}

	if urlIns.Scheme != "file" {
		return "", errors.New("cannot convert non-file URI")
	}
	filePath, err := url.QueryUnescape(urlIns.Path)

	if err != nil {
		return "", err
	}

	if strings.Contains(filePath, ":") {
		filePath = strings.TrimPrefix(filePath, "/")
		filePath = strings.Replace(filePath, "/", "\\", -1)
	}

	return filePath, nil
}

// PathToURI converts file path to URI
func PathToURI(path string) string {
	urlIns := new(url.URL)
	path = strings.Replace(strings.Trim(path, "/"), "\\", "/", -1)
	parts := strings.Split(path, "/")
	first, parts := parts[0], parts[1:]

	if !strings.HasSuffix(first, ":") {
		first = url.QueryEscape(first)
	}

	tempParts := parts[:0]
	for _, part := range parts {
		tempParts = append(tempParts, url.QueryEscape(part))
	}
	parts = append([]string{first}, tempParts...)

	urlIns.Scheme = "file"
	urlIns.Path = "/" + strings.Join(parts, "/")

	return urlIns.String()
}

// GetURIID shortens the URI but still readable
func GetURIID(uri string) string {
	u, err := url.Parse(uri)
	h := md5.New()
	io.WriteString(h, uri)
	hash := hex.EncodeToString(h.Sum(nil))
	// Fallback to MD5 only if uri cannot be parsed
	if err != nil {
		return hash
	}
	parts := strings.Split(u.Path, "/")
	if parts[len(parts)-1] == "" {
		parts = parts[:len(parts)-1]
	}
	parts, lastPart := parts[:len(parts)-1], parts[len(parts)-1]
	result := lastPart + "-"
	if len(parts) > 0 {
		for _, part := range parts[0 : len(parts)-1] {
			r := []rune(part)
			if len(r) == 0 {
				continue
			}
			result += string(r[0])
		}
	}
	result += "-" + hash[:8]
	return result
}

// CanonicaliseURI canonicalises a child URI by shorten it assuming that it is
// prefixed with parent, otherwise the original URI is returned
func CanonicaliseURI(parent string, child string) string {
	return strings.TrimPrefix(child, parent)
}

// URIFromCanonicalURI converts canonical URI back to full URI, if the canonical URI
// is a full URI which is prefixed with parent then it is returned
func URIFromCanonicalURI(parent string, canonicalURI string) string {
	if strings.HasPrefix(canonicalURI, parent) {
		return canonicalURI
	}
	return parent + canonicalURI
}

// IsURINavigatable checks if the given URI is navigatable
func IsURINavigatable(uri string) bool {
	// TODO: Currently, the only navigatable URI is file
	// but this is not true because there are also other navigatable URIs
	// e.g. FTP, HTTP or other storage providers
	return strings.HasPrefix(uri, "file://")
}

func DecodeURIFromQuery(uri string) (string, error) {
	uri, err := url.QueryUnescape(uri)
	if err != nil {
		return "", fmt.Errorf("workspaceStore.getStore cannot QueryUnescape %s, err: %v", uri, err)
	}
	u, err := url.Parse(uri)
	if err != nil {
		return "", fmt.Errorf("workspaceStore.getStore cannot Parse %s, err: %v", uri, err)
	}
	uri = u.String()
	return uri, nil
}
