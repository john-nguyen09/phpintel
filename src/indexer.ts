import * as fs from "fs";
import * as path from "path";
import { promisify } from "util";
import { PhpFile } from "./phpFile";
import { TreeAnalyser } from "./treeAnalyser";

const readdirAsync = promisify(fs.readdir);
const statAsync = promisify(fs.stat);

export interface FileInfo {
    filePath: fs.PathLike;
    stats: fs.Stats;
}

export namespace FileInfo {
    export async function create(path: fs.PathLike): Promise<FileInfo> {
        return {
            filePath: path,
            stats: await statAsync(path)
        }
    }
}

export namespace Indexer {
    export async function indexWorkspace(directory: fs.PathLike): Promise<void> {
        let directories: fs.PathLike[] = [
            directory
        ];
        const promises: Promise<void>[] = [];

        while (directories.length > 0) {
            let dir = directories.shift();
            if (dir === undefined) {
                continue;
            }
            let files = await readdirAsync(dir);

            for (let file of files) {
                const fileInfo = await FileInfo.create(path.join(dir.toString(), file));

                if (fileInfo.stats.isDirectory()) {
                    directories.push(fileInfo.filePath);

                    continue;
                }

                if (file.endsWith('.php')) {
                    promises.push(PhpFile.create(fileInfo).then((phpFile) => {
                        TreeAnalyser.analyse(phpFile);
                    }));
                }
            }
        }

        await Promise.all(promises);
    }
}