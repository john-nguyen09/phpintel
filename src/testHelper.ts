import { PhpDocument } from "./symbol/phpDocument";
import { pathToUri } from "./util/uri";
import * as fs from "fs";
import { SymbolParser } from "./symbol/symbolParser";
import { Traverser } from "./traverser";
import { TreeNode } from "./util/parseTree";
import { Parser, Phrase, phraseTypeToString, tokenTypeToString } from "php7parser";
import * as path from "path";
import { inspect } from "util";

export function indexFiles(filePaths: string[]): PhpDocument[] {
    let phpDocs: PhpDocument[] = [];
    let treeTraverser = new Traverser();

    for (let filePath of filePaths) {
        const fileUri = pathToUri(filePath);
        const fileContent = fs.readFileSync(filePath).toString();
        const symbolParser = new SymbolParser(new PhpDocument(fileUri, fileContent));
        const parseTree = Parser.parse(fileContent);
        
        treeTraverser.traverse(parseTree, [symbolParser]);

        phpDocs.push(symbolParser.getTree());
    }

    return phpDocs;
}

export function getCaseDir(): string {
    return path.resolve(__dirname, "..", "case");
}

export function getDebugDir(): string {
    return path.resolve(__dirname, '..', 'debug');
}

export function dumpToDebug(name: string, object: any, depth?: number): void {
    fs.writeFile(
        path.resolve(__dirname, '..', 'debug', name),
        inspect(object, {
            depth: depth
        }),
        (err) => {
            console.error(err);
        }
    );
}

export function dumpAstToDebug(name: string, parseTree: Phrase): void {
    fs.writeFile(
        path.resolve(__dirname, '..', 'debug', name),
        JSON.stringify(parseTree, (key, value) => {
            if (key == 'modeStack') {
                return undefined;
            }

            if (key == 'phraseType') {
                return phraseTypeToString(value);
            }

            if (key == 'tokenType') {
                return tokenTypeToString(value);
            }

            return value;
        }, 2),
        (err) => {
            if (err) {
                console.log(err);
            }
        }
    );
}