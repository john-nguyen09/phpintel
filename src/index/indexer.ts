import * as fs from "fs";
import * as path from "path";
import { promisify } from "util";
import { pathToUri, uriToPath } from "../util/uri";
import { SymbolParser } from "../symbol/symbolParser";
import { PhpDocument } from "../symbol/phpDocument";
import { injectable } from "inversify";
import { Traverser } from "../traverser";
import { ClassTable } from "../storage/table/class";
import { ClassConstantTable } from "../storage/table/classConstant";
import { ConstantTable } from "../storage/table/constant";
import { FunctionTable } from "../storage/table/function";
import { MethodTable } from "../storage/table/method";
import { PropertyTable } from "../storage/table/property";
import { PhpDocumentTable } from "../storage/table/phpDoc";
import { GlobalVariableTable } from "../storage/table/globalVariable";

const readdirAsync = promisify(fs.readdir);
const readFileAsync = promisify(fs.readFile);
const statAsync = promisify(fs.stat);

export interface PhpFileInfo {
    filePath: string;
    fstats: fs.Stats;
}

export namespace PhpFileInfo {
    export async function createFileInfo(filePath: string): Promise<PhpFileInfo> {
        return {
            filePath: filePath,
            fstats: await statAsync(filePath)
        };
    }
}

@injectable()
export class Indexer {
    private openedUri: Set<string> = new Set<string>();

    constructor(
        private treeTraverser: Traverser,
        private phpDocTable: PhpDocumentTable,
        private classTable: ClassTable,
        private classConstantTable: ClassConstantTable,
        private constantTable: ConstantTable,
        private functionTable: FunctionTable,
        private methodTable: MethodTable,
        private propertyTable: PropertyTable,
        private globalVariableTable: GlobalVariableTable
    ) { }

    async getOrCreatePhpDoc(uri: string): Promise<PhpDocument> {
        let phpDoc = await this.phpDocTable.get(uri);

        if (phpDoc === null) {
            let fileContent = (await readFileAsync(uriToPath(uri))).toString('utf-8');
            phpDoc = new PhpDocument(uri, fileContent);
        }

        return phpDoc;
    }

    async syncFileSystem(fileInfo: PhpFileInfo): Promise<void> {
        let fileUri = pathToUri(fileInfo.filePath);
        const fileModifiedTime = Math.round(fileInfo.fstats.mtime.getTime() / 1000);

        let phpDoc = await this.getOrCreatePhpDoc(fileUri);
        if (phpDoc.modifiedTime !== fileModifiedTime) {
            phpDoc.modifiedTime = fileModifiedTime;
            await this.indexFile(phpDoc);
        }
    }

    async indexFile(phpDoc: PhpDocument): Promise<void> {
        let symbolParser = new SymbolParser(phpDoc);

        this.treeTraverser.traverse(phpDoc.getTree(), [symbolParser]);
        await this.indexPhpDocument(symbolParser.getPhpDoc());
    }

    async indexWorkspace(directory: string): Promise<void> {
        let directories: string[] = [
            directory
        ];
        const promises: Promise<void>[] = [];

        while (directories.length > 0) {
            let dir = directories.shift();
            if (dir === undefined) {
                continue;
            }
            let files = await readdirAsync(dir);

            for (let file of files) {
                const fileInfo = await PhpFileInfo.createFileInfo(path.join(dir, file));

                if (file.endsWith('.php')) {
                    promises.push(this.syncFileSystem(fileInfo));
                } else if (fileInfo.fstats.isDirectory()) {
                    directories.push(fileInfo.filePath);
                }
            }
        }

        await Promise.all(promises);
    }

    public async open(uri: string): Promise<void> {
        this.openedUri.add(uri);
        const phpDoc = await this.getOrCreatePhpDoc(uri);
        await this.indexFile(phpDoc);
    }

    public close(uri: string): void {
        this.openedUri.delete(uri);
    }

    private async removeSymbolsByDoc(uri: string) {
        return Promise.all([
            this.classTable.removeByDoc(uri),
            this.classConstantTable.removeByDoc(uri),
            this.constantTable.removeByDoc(uri),
            this.functionTable.removeByDoc(uri),
            this.methodTable.removeByDoc(uri),
            this.propertyTable.removeByDoc(uri),
            this.globalVariableTable.removeByDoc(uri),
        ]);
    }

    private async indexPhpDocument(doc: PhpDocument): Promise<void> {
        await this.removeSymbolsByDoc(doc.uri);

        const promises: Promise<void | void[] | void[][]>[] = [];

        for (const theClass of doc.classes) {
            promises.push(this.classTable.put(doc, theClass));
        }

        for (const classConstant of doc.classConstants) {
            promises.push(this.classConstantTable.put(doc, classConstant));
        }

        for (const constant of doc.constants) {
            promises.push(this.constantTable.put(doc, constant));
        }

        for (const func of doc.functions) {
            promises.push(this.functionTable.put(doc, func));
        }

        for (const method of doc.methods) {
            promises.push(this.methodTable.put(doc, method));
        }

        for (const property of doc.properties) {
            promises.push(this.propertyTable.put(doc, property));
        }

        for (const globalVariable of doc.globalVariables) {
            globalVariable.assignExtraTypeForVariables();
            promises.push(this.globalVariableTable.put(doc, globalVariable));
        }

        await Promise.all(promises);
        await this.phpDocTable.put(doc, this.openedUri.has(doc.uri));
    }
}