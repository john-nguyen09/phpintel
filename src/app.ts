import "reflect-metadata";
import { LevelDatasource } from "./storage/db";
import { LogWriter } from "./service/logWriter";
import { Hasher } from "./service/hasher";
import { Container, interfaces } from "inversify";
import { Indexer } from "./index/indexer";
import { ClassTable } from "./storage/table/class";
import { ClassConstantTable } from "./storage/table/classConstant";
import { ConstantTable } from "./storage/table/constant";
import { FunctionTable } from "./storage/table/function";
import { MethodTable } from "./storage/table/method";
import { PropertyTable } from "./storage/table/property";
import { Traverser } from "./traverser";
import { ReferenceTable } from "./storage/table/referenceTable";
import { PhpDocumentTable } from "./storage/table/phpDoc";
import { IConnection } from "vscode-languageserver";

export class Application {
    protected container: Container = new Container();

    constructor(storageLocation: string, connection?: IConnection) {
        this.container.bind<IConnection | undefined>('IConnection').toConstantValue(connection);

        this.initBind();
        this.initStorage(storageLocation);
    }

    public getContainer(): Container {
        return this.container;
    }

    protected initStorage(location: string) {
        let datasource = new LevelDatasource(location);
        this.container.bind<LevelDatasource>(LevelDatasource).toConstantValue(datasource);

        // Tables
        this.container.bind<ClassTable>(ClassTable).toSelf();
        this.container.bind<ClassConstantTable>(ClassConstantTable).toSelf();
        this.container.bind<ConstantTable>(ConstantTable).toSelf();
        this.container.bind<FunctionTable>(FunctionTable).toSelf();
        this.container.bind<MethodTable>(MethodTable).toSelf();
        this.container.bind<PropertyTable>(PropertyTable).toSelf();
        this.container.bind<PhpDocumentTable>(PhpDocumentTable).toSelf();
        this.container.bind<ReferenceTable>(ReferenceTable).toSelf();
    }

    protected initBind() {
        this.container.bind<LogWriter>(LogWriter).toSelf().inSingletonScope();
        this.container.bind<Hasher>(Hasher).toSelf().inSingletonScope();
        this.container.bind<Indexer>(Indexer).toSelf().inSingletonScope();
        this.container.bind<Traverser>(Traverser).toSelf().inSingletonScope();
    }
}

export namespace App {
    let app: Application;

    export function init(storageLocation: string, connection?: IConnection) {
        app = new Application(storageLocation, connection);
    }

    export function get<T>(
        identifier: string | symbol | interfaces.Newable<T> | interfaces.Abstract<T>
    ): T {
        return app.getContainer().get(identifier);
    }
}
