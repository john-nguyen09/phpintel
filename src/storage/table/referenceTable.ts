import { Reference } from "../../symbol/symbol";
import { DbStore, LevelDatasource, SubStore } from "../db";
import { Serializer } from "../serializer";
import { TypeName } from "../../type/name";
import { TypeComposite } from "../../type/composite";
import { injectable } from "inversify";
import { App } from "../../app";
import { LogWriter } from "../../service/logWriter";

@injectable()
export class ReferenceTable {
    private db: DbStore;

    constructor(level: LevelDatasource) {
        this.db = new SubStore(level, {
            name: 'reference',
            version: 1
        }, ReferenceEncoding);
    }

    async put(reference: Reference) {
        if (reference.location.isEmpty) {
            return;
        }

        let serializer = new Serializer();
        serializer.writeInt32(reference.location.range.start.offset);
        serializer.writeInt32(reference.location.range.end.offset);

        let key = Buffer.concat([
            Buffer.from(reference.location.uri),
            serializer.getBuffer()
        ]);

        return this.db.put(key, reference);
    }

    async removeByDoc(uri: string) {
        const db = this.db;

        return new Promise<void>((resolve, reject) => {
            db.prefixSearch(uri)
                .on('data', (data) => {
                    db.del(data.key);
                })
                .on('error', (err) => {
                    reject(err);
                })
                .on('end', () => {
                    resolve();
                });
        });
    }

    async findAt(uri: string, offset: number): Promise<Reference | null> {
        const db = this.db;
        // const logger = App.get<LogWriter>(LogWriter);

        return new Promise<Reference | null>((resolve, reject) => {
            let serializer = new Serializer();
            serializer.writeInt32(offset);

            let uriBuffer = Buffer.from(uri);

            let key = Buffer.concat([
                uriBuffer,
                serializer.getBuffer()
            ]);
            let iterator = db.iterator({
                lt: key,
                gte: uriBuffer
            });

            const processRef = (
                err: Error | null,
                key?: string | Buffer,
                ref?: Reference
            ): void => {
                if (err) {
                    iterator.end(() => { });

                    return reject(err);
                }

                // End of stream reached
                if (key == undefined || ref == undefined) {
                    iterator.end(() => {
                        resolve(null);
                    });
                    return;
                }

                // logger.info(JSON.stringify(ref));
                // logger.info(ref.location.range.end.offset.toString());
                // logger.info(offset.toString());
                // logger.info(JSON.stringify(ref.location.range.end.offset >= offset));

                if (ref.location.range.end.offset >= offset) {
                    iterator.end(() => {
                        resolve(ref);
                    });

                    return;
                }

                iterator.next(processRef);
            }

            iterator.next(processRef);
        });
    }
}

enum TypeKind {
    TYPE_NAME = 1,
    TYPE_COMPOSITE = 2
};

const ReferenceEncoding = {
    type: 'reference-encoding',
    encode(ref: Reference): Buffer {
        let serializer = new Serializer();

        if (ref.type instanceof TypeName) {
            serializer.writeInt32(TypeKind.TYPE_NAME);
            serializer.writeTypeName(ref.type);
        } else if (ref.type instanceof TypeComposite) {
            serializer.writeInt32(TypeKind.TYPE_COMPOSITE);
            serializer.writeTypeComposite(ref.type);
        }

        serializer.writeLocation(ref.location);

        return serializer.getBuffer();
    },
    decode(buffer: Buffer): Reference | null {
        if (buffer.length === 0) {
            return null;
        }

        let serializer = new Serializer(buffer);
        let type: TypeName | TypeComposite = new TypeName('');
        let typeKind: TypeKind = serializer.readInt32();

        if (typeKind == TypeKind.TYPE_NAME) {
            type = serializer.readTypeName() || new TypeName('');
        } else if (typeKind === TypeKind.TYPE_COMPOSITE) {
            type = serializer.readTypeComposite();
        }
        
        return {
            type: type,
            location: serializer.readLocation()
        } as Reference;
    }
} as Level.Encoding;