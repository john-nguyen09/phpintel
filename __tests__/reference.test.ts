import { App } from '../src/app';
import { Indexer } from '../src/index/indexer';
import { getCaseDir, getDebugDir } from "../src/testHelper";
import * as path from "path";
import { ReferenceTable } from '../src/storage/table/referenceTable';
import { pathToUri } from '../src/util/uri';

describe('Testing functions around references', () => {
    it('should return the reference at the cursor', async () => {
        App.init(path.join(getDebugDir(), 'storage'));

        const indexer = App.get<Indexer>(Indexer);
        const caseDir = getCaseDir();
        const refTable = App.get<ReferenceTable>(ReferenceTable);
        const refTestFile = path.join(caseDir, 'reference', 'references.php');

        await indexer.indexFile(path.join(caseDir, 'global_symbols.php'));
        await indexer.indexFile(path.join(caseDir, 'function_declare.php'));
        await indexer.indexFile(refTestFile);

        let refTestUri = pathToUri(refTestFile);
        let refs = [
            await refTable.findAt(refTestUri, 7),
            await refTable.findAt(refTestUri, 14),
            await refTable.findAt(refTestUri, 10),
            await refTable.findAt(refTestUri, 37),
            await refTable.findAt(refTestUri, 51),
            await refTable.findAt(refTestUri, 42),
        ];

        expect(refs).toMatchSnapshot();
    });
});