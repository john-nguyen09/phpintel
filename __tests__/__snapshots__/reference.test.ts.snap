// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing functions around references should return the reference at the cursor 1`] = `
Array [
  Object {
    "location": Location {
      "range": Range {
        "end": Position {
          "character": 7,
          "line": 1,
          "offset": 14,
        },
        "start": Position {
          "character": 0,
          "line": 1,
          "offset": 7,
        },
      },
      "uri": "file:///C:/Users/johnn/Development/phpintel/case/reference/references.php",
    },
    "refKind": 2,
    "type": TypeComposite {
      "_types": Array [
        TypeName {
          "name": "TestClass",
        },
      ],
    },
  },
  Object {
    "location": Location {
      "range": Range {
        "end": Position {
          "character": 7,
          "line": 1,
          "offset": 14,
        },
        "start": Position {
          "character": 0,
          "line": 1,
          "offset": 7,
        },
      },
      "uri": "file:///C:/Users/johnn/Development/phpintel/case/reference/references.php",
    },
    "refKind": 2,
    "type": TypeComposite {
      "_types": Array [
        TypeName {
          "name": "TestClass",
        },
      ],
    },
  },
  Object {
    "location": Location {
      "range": Range {
        "end": Position {
          "character": 7,
          "line": 1,
          "offset": 14,
        },
        "start": Position {
          "character": 0,
          "line": 1,
          "offset": 7,
        },
      },
      "uri": "file:///C:/Users/johnn/Development/phpintel/case/reference/references.php",
    },
    "refKind": 2,
    "type": TypeComposite {
      "_types": Array [
        TypeName {
          "name": "TestClass",
        },
      ],
    },
  },
  Object {
    "location": Location {
      "range": Range {
        "end": Position {
          "character": 16,
          "line": 3,
          "offset": 53,
        },
        "start": Position {
          "character": 0,
          "line": 3,
          "offset": 37,
        },
      },
      "uri": "file:///C:/Users/johnn/Development/phpintel/case/reference/references.php",
    },
    "refKind": 1,
    "type": TypeName {
      "name": "test_function1",
    },
  },
  Object {
    "location": Location {
      "range": Range {
        "end": Position {
          "character": 16,
          "line": 3,
          "offset": 53,
        },
        "start": Position {
          "character": 0,
          "line": 3,
          "offset": 37,
        },
      },
      "uri": "file:///C:/Users/johnn/Development/phpintel/case/reference/references.php",
    },
    "refKind": 1,
    "type": TypeName {
      "name": "test_function1",
    },
  },
  Object {
    "location": Location {
      "range": Range {
        "end": Position {
          "character": 16,
          "line": 3,
          "offset": 53,
        },
        "start": Position {
          "character": 0,
          "line": 3,
          "offset": 37,
        },
      },
      "uri": "file:///C:/Users/johnn/Development/phpintel/case/reference/references.php",
    },
    "refKind": 1,
    "type": TypeName {
      "name": "test_function1",
    },
  },
]
`;
